### PR TITLE
Parallel RF

### PR DESCRIPTION
**Pull request by Saurabh7**
WIP

### CODE REVIEW

File: src/shogun/features/DenseFeatures.cpp [hunk 1]
```diff
@@ -636,6 +636,20 @@ CFeatures* CDenseFeatures<ST>::copy_dimension_subset(SGVector<index_t> dims)
 	return result;
 }
 
+template<class ST>
+CFeatures* CDenseFeatures<ST>::shallow_subset_copy()
+{
+	CFeatures* shallow_copy_features=nullptr;
+	
+	SG_SDEBUG("Using underlying feature matrix with %d dimensions and %d feature vectors!\n", num_features, num_vectors);
+	SGMatrix<ST> shallow_copy_matrix = SGMatrix<ST>(feature_matrix.matrix, num_features, num_vectors, false);
```
Comment (karlnapf) [inline src/shogun/features/DenseFeatures.cpp]: You can use the copy constructor of SGMatrix

File: src/shogun/features/DenseFeatures.cpp [hunk 2]
```diff
@@ -636,6 +636,20 @@ CFeatures* CDenseFeatures<ST>::copy_dimension_subset(SGVector<index_t> dims)
...
+	shallow_copy_features=new CDenseFeatures<ST>(shallow_copy_matrix);
+	if (m_subset_stack->has_subsets())
```
Comment (karlnapf) [inline src/shogun/features/DenseFeatures.cpp]: seems good like this

File: src/shogun/labels/DenseLabels.cpp [hunk 3]
```diff
@@ -203,3 +203,4 @@ int32_t CDenseLabels::get_num_labels() const
 	return m_subset_stack->has_subsets()
 			? m_subset_stack->get_size() : m_labels.vlen;
 }
+
```
Comment (karlnapf) [inline src/shogun/labels/DenseLabels.cpp]: please dont do such whitespace changes

File: src/shogun/labels/MulticlassLabels.cpp [hunk 4]
```diff
@@ -134,3 +134,16 @@ int32_t CMulticlassLabels::get_num_classes()
 	SGVector<float64_t> unique=get_unique_labels();
 	return unique.vlen;
 }
+
+CLabels* CMulticlassLabels::shallow_subset_copy()
+{
+	CLabels* shallow_copy_labels=nullptr;
+	SGVector<float64_t> shallow_copy_vector = SGVector<float64_t>(m_labels.vector, m_labels.size(), false);
```
Comment (karlnapf) [inline src/shogun/labels/MulticlassLabels.cpp]: copy constr

File: src/shogun/labels/MulticlassLabels.cpp [hunk 5]
```diff
@@ -134,3 +134,16 @@ int32_t CMulticlassLabels::get_num_classes()
 	SGVector<float64_t> unique=get_unique_labels();
 	return unique.vlen;
 }
+
+CLabels* CMulticlassLabels::shallow_subset_copy()
```
Comment (karlnapf) [inline src/shogun/labels/MulticlassLabels.cpp]: BTW for both of those methods: You either need to hide them from SWIG or put a `%newobject` directive in the interface file

File: src/shogun/multiclass/tree/CARTree.cpp [hunk 6]
```diff
@@ -603,8 +603,6 @@ int32_t CCARTree::compute_best_attribute(const SGMatrix<float64_t>& mat, const S
 	{
 		SGVector<float64_t> feats(num_vecs);
 		SGVector<index_t> sorted_args(num_vecs);
```
Comment (karlnapf) [inline src/shogun/multiclass/tree/CARTree.cpp]: are the changes in this file related to parallel? If not, pls separate pr

File: tests/unit/features/DenseFeatures_unittest.cc [hunk 7]
```diff
@@ -119,3 +119,33 @@ TEST(DenseFeaturesTest, copy_dimension_subset_with_subsets)
 	SG_UNREF(features);
 	SG_UNREF(f_reduced);
 }
+
+TEST(DenseFeaturesTest, shallow_copy_subset_data)
+{
+	index_t dim=5;
+	index_t n=10;
+
+	SGMatrix<float64_t> data(dim, n);
```
Comment (karlnapf) [inline tests/unit/features/DenseFeatures_unittest.cc]: Use iota on data.matrix. One line of code

File: tests/unit/features/DenseFeatures_unittest.cc [hunk 8]
```diff
@@ -119,3 +119,33 @@ TEST(DenseFeaturesTest, copy_dimension_subset_with_subsets)
...
+	for (index_t i=0; i<dim; ++i)
+		for (index_t j=0; j<n; ++j)	
+			data(i,j)=CMath::random(0, n-1);
+	
+	SGVector<index_t> inds(n/2);
```
Comment (karlnapf) [inline tests/unit/features/DenseFeatures_unittest.cc]: we have a permutation methods
range_fill (or iota) and then permute

File: tests/unit/features/DenseFeatures_unittest.cc [hunk 9]
```diff
@@ -119,3 +119,33 @@ TEST(DenseFeaturesTest, copy_dimension_subset_with_subsets)
 	SG_UNREF(features);
 	SG_UNREF(f_reduced);
 }
+
+TEST(DenseFeaturesTest, shallow_copy_subset_data)
```
Comment (karlnapf) [inline tests/unit/features/DenseFeatures_unittest.cc]: Need more tests.
Pls add some tests where you ensure that the objects can be used within threads. 
I.e. modifying the shallow_subset copy doesnt change another one...
etc

**Comment by karlnapf** [comment]
So this works?
Whats the gain with increasing number of cores?

**Comment by Saurabh7** [comment]
Yes, it works.
Although calculating oob error is leading to change in values for test. 
I suspected this is due to chaning of order in which we append oob indices (multi thread) to dynamic array. So I made it critical for the thread, but still there are some changes.

outputs: https://gist.github.com/Saurabh7/40889d51c856f73c33337408eb3a7fd1

**Comment by karlnapf** [comment]
I guess you can force a single thread for the result tests.
Multithread tests can be done on the feature subset shallow copy alone, independently of rf

**Comment by Saurabh7** [comment]
Ok so few things for this:
currently this Mock test which uses mock labels fails: https://github.com/shogun-toolbox/shogun/blob/develop/tests/unit/multiclass/BaggingMachine_unittest.cc#L35. because this shallow_subset_copy method is not implemented in Labels base class.

Regarding If objects are safe to use in other algorithms, `get_feature_matrix()` returns a copy if subset are present so thats not a problem.
other access methods like `get_feature_vector` and `set_feature_vector` will change the data if written using it.

**Comment by Saurabh7** [comment]
Also I found the changes in oob error were because, randomly generated indices in multithreaded version keep chaning even thought I set seed in sg_rand.

File: src/shogun/labels/MulticlassLabels.cpp (hunk 5 above)
Comment (karlnapf) [inline src/shogun/labels/MulticlassLabels.cpp]: this is still a blocker

**Comment by karlnapf** [comment]
And merge :)

**Comment by karlnapf** [comment]
ah crap, it wasnt squashed ;)
Can you always make sure you squash your commits before asking to merge them?
I know I should check, but there is a lot of stuff going on atm

**Comment by Saurabh7** [comment]
@karlnapf ah ok sorry :) Next time I will put an explicit comment when final.

**Pull request merged**